finality-aleph/src/lib.rs

```diff
@@ -218,36 +220,64 @@ where
 {
 }
 
+/// The identifier of a block, the least amount of knowledge we can have about a block.
+pub trait BlockIdentifier: Clone + Hash + Debug + Eq + Codec + Send + Sync + 'static {
+    /// The block number, useful when reasoning about hopeless forks.
+    fn number(&self) -> BlockNumber;
+}
+
 type Hasher = abft::HashWrapper<BlakeTwo256>;
 
-#[derive(PartialEq, Eq, Clone, Debug, Hash)]
-pub struct HashNum<H, N> {
-    hash: H,
-    num: N,
+#[derive(PartialEq, Eq, Clone, Debug, Encode, Decode)]
+pub struct HashNum<H: Header> {
+    hash: H::Hash,
+    num: H::Number,
 }
 
-impl<H> HashNum<H, BlockNumber> {
-    fn new(hash: H, num: BlockNumber) -> Self {
+impl<H: Header<Number = BlockNumber>> HashNum<H> {
+    fn new(hash: H::Hash, num: BlockNumber) -> Self {
         HashNum { hash, num }
     }
 }
 
-impl<H> From<(H, BlockNumber)> for HashNum<H, BlockNumber> {
-    fn from(pair: (H, BlockNumber)) -> Self {
+impl<H: Header<Number = BlockNumber>> From<(H::Hash, BlockNumber)> for HashNum<H> {
+    fn from(pair: (H::Hash, BlockNumber)) -> Self {
         HashNum::new(pair.0, pair.1)
     }
 }
 
-pub type BlockHashNum<B> = HashNum<<B as Block>::Hash, NumberFor<B>>;
+impl<SH: Header> Hash for HashNum<SH> {
+    fn hash<H>(&self, state: &mut H)
+    where
+        H: std::hash::Hasher,
+    {
+        self.hash.hash(state);
+        self.num.hash(state);
+    }
+}
+
+pub type BlockHashNum<B> = HashNum<<B as Block>::Header>;
+pub type IdentifierFor<B> = HashNum<<B as Block>::Header>;
```

Comment:
nit:
```suggestion
pub type IdentifierFor<B> = BlockHashNum<B>
```

Reply:
actually why we need both BlockHashNum and IdentifierFor?

Reply:
`BlockHashNum` appears in around 72 places in the code. I thought it would be better to unify those two in another PR then to make this one even bigger. Also in this other PR there will be changes like unifying `BlockId` from sync with `HashNum` (using `BlockId` struct everywhere). They are almost the same but this includes changing `.num` to `.number` and also renaming some variables. A lot of changes.

Reply:
If you want me to I can change this to `pub type IdentifierFor<B> = BlockHashNum<B>` but tbh it will be changed anyway in another PR.

Reply:
it is fine llike it is